x/concentrated-liquidity/lp.go

```diff
@@ -141,28 +141,13 @@ func (k Keeper) withdrawPosition(ctx sdk.Context, owner sdk.AccAddress, position
 
 	// If underlying lock exists in state, validate unlocked conditions are met before withdrawing liquidity.
 	// If unlocked conditions are met, remove the link between the position and the underlying lock.
-	positionHasUnderlyingLock, err := k.positionHasUnderlyingLockInState(ctx, positionId)
+	positionHasActiveUnderlyingLock, lockId, err := k.PositionHasActiveUnderlyingLock(ctx, positionId)
 	if err != nil {
 		return sdk.Int{}, sdk.Int{}, err
 	}
-	if positionHasUnderlyingLock {
-		lockId, err := k.GetLockIdFromPositionId(ctx, positionId)
-		if err != nil {
-			return sdk.Int{}, sdk.Int{}, err
-		}
-
-		// Check if the underlying lock is mature.
-		lockIsMature, err := k.isLockMature(ctx, lockId)
-		if err != nil {
-			return sdk.Int{}, sdk.Int{}, err
-		}
-		if lockIsMature {
-			// Remove the link between the position and the underlying lock since the lock is mature.
-			k.RemovePositionIdToLock(ctx, positionId, lockId)
-		} else {
-			// Lock is not mature, return error.
-			return sdk.Int{}, sdk.Int{}, types.LockNotMatureError{PositionId: position.PositionId, LockId: lockId}
-		}
+	if positionHasActiveUnderlyingLock {
+		// Lock is not mature, return error.
+		return sdk.Int{}, sdk.Int{}, types.LockNotMatureError{PositionId: position.PositionId, LockId: lockId}
```

Comment:
Can you clarify what the purpose of this change was? It seems as though the new code does not call `RemovePositionIdToLock` for mature locks – I'm assuming this is intended?

Reply:
This is because I refactored PositionHasActiveUnderlyingLock to lazily update whenever fetching the underlying lock